### PR TITLE
(WIP)(PUP-6374) Minitar verify module tarballs

### PR DESCRIPTION
**Pull request by Iristyle**
- Previously it was possible, depending on platform, to create a
  .tar.gz that won't properly extract with Minitar on puppet versions
  prior to 3.8.7 or 4.4.0.  This has particularly caused problems
  with the DSC module on Windows.
  
  While the bug was fixed in #4740 /
  f888fb09fbef5f7c74ef799b3877ec6f7843a409, these older versions may
  still be unable to install modules as a result on platforms that
  rely on minitar - namely Windows, AIX and Solaris
- Provide some ability to catch these modules as they're packed on
  platforms with a native tar.  This should primarly be OSX, where
  bsdtar is the norm.  The use of gnu tar typically creates archives
  compatible with the old version of Puppet that does not contain
  the minitar fix.

### CODE REVIEW

File: lib/puppet/module_tool/applications/builder.rb [hunk 1]
```diff
@@ -40,6 +41,22 @@ def archive_file
...
+          require 'archive/tar/minitar'
+        rescue
+          return true
+        end
+
+        # minitar present, so assume invalid
+        files = Puppet::ModuleTool::Tar::Mini.new.entries(archive_file)
+        valid = files.all? { |f| f.full_name == f.name }
+        if !valid
+          raise Puppet::ModuleTool::Errors::ModuleToolError, "Archive contains invalid entries - may fail minitar extraction on Windows."
```
Comment (Iristyle) [inline lib/puppet/module_tool/applications/builder.rb]: I really want this to be a warning locally, but an error in CI.  Perhaps this needs to be controlled by a global setting.

File: lib/puppet/module_tool/applications/builder.rb [hunk 2]
```diff
@@ -40,6 +41,22 @@ def archive_file
         File.join(@pkg_path, "#{metadata.release_name}.tar.gz")
       end
 
+      def minitar_verify
+        begin
+          # assume valid if minitar inaccessible (i.e. immediate rescue loading gem)
+          require 'archive/tar/minitar'
```
Comment (Iristyle) [inline lib/puppet/module_tool/applications/builder.rb]: The general idea here is that modules would opt into this behavior by adding `minitar` to their `Gemfile`.

This wouldn't take effect under normal circumstances.

File: lib/puppet/module_tool/applications/builder.rb [hunk 3]
```diff
@@ -40,6 +41,22 @@ def archive_file
...
+        rescue
+          return true
+        end
+
+        # minitar present, so assume invalid
+        files = Puppet::ModuleTool::Tar::Mini.new.entries(archive_file)
+        valid = files.all? { |f| f.full_name == f.name }
```
Comment (Iristyle) [inline lib/puppet/module_tool/applications/builder.rb]: Not even positive this comparison is accurate - on OSX, when run with DSC module, the entries contain a bunch of oddly encoded strings.  Not sure if minitar just doesn't work right on OSX?

**Comment by MosesMendoza** [comment]
This PR has been open for some time and unfortunately it seems like we're not going to be able to pick it up any time soon. I'm going to close it for now, but we can re-open if/when PUP-6374 is prioritized for next steps.

**Pull request closed without merging**